pkg/manager/member/tidbcluster_status_manager.go

```diff
@@ -14,37 +14,180 @@
+func NewTidbClusterStatusManager(kubeCli kubernetes.Interface, cli versioned.Interface) *TidbClusterStatusManager {
 	return &TidbClusterStatusManager{
-		cli: cli,
+		cli:       cli,
+		pdControl: pdapi.NewDefaultPDControl(kubeCli),
 	}
 }
 
 func (tcsm *TidbClusterStatusManager) Sync(tc *v1alpha1.TidbCluster) error {
+	return tcsm.syncTidbMonitorRefAndKey(tc)
+}
+
+func (tcsm *TidbClusterStatusManager) syncTidbMonitorRefAndKey(tc *v1alpha1.TidbCluster) error {
+	tm, err := tcsm.syncTidbMonitorRef(tc)
+	if err != nil {
+		return err
+	}
+	return tcsm.syncDashboardMetricStorage(tc, tm)
+}
+
+func (tcsm *TidbClusterStatusManager) syncTidbMonitorRef(tc *v1alpha1.TidbCluster) (*v1alpha1.TidbMonitor, error) {
+	if tc.Status.Monitor == nil {
+		return nil, nil
+	}
+	tmRef := tc.Status.Monitor
+	tm, err := tcsm.cli.PingcapV1alpha1().TidbMonitors(tmRef.Namespace).Get(tmRef.Name, metav1.GetOptions{})
+	if err != nil {
+		if errors.IsNotFound(err) {
+			tc.Status.Monitor = nil
+			err = nil
+		}
+		return nil, err
+	}
+	tcRef := tm.Spec.Clusters
+	if len(tcRef) < 1 {
+		tc.Status.Monitor = nil
+		return nil, nil
+	}
+	if len(tcRef[0].Namespace) < 1 {
+		tcRef[0].Namespace = tm.Namespace
+	}
+	if tcRef[0].Name != tc.Name || tcRef[0].Namespace != tc.Namespace {
+		tc.Status.Monitor = nil
+		return nil, nil
+	}
+
+	return tm, nil
+}
+
+func (tcsm *TidbClusterStatusManager) syncDashboardMetricStorage(tc *v1alpha1.TidbCluster, tm *v1alpha1.TidbMonitor) error {
+	pdEtcdClient, err := tcsm.pdControl.GetPDEtcdClient(pdapi.Namespace(tc.Namespace), tc.Name, tc.IsTLSClusterEnabled())
+	if err != nil {
+		return err
+	}
+	var prometheusExist bool
+	var grafanaExist bool
 	if tc.Status.Monitor != nil {
-		tmRef := tc.Status.Monitor
-		_, err := tcsm.cli.PingcapV1alpha1().TidbMonitors(tmRef.Namespace).Get(tmRef.Name, metav1.GetOptions{})
+		prometheusExist = true
+		if tm.Spec.Grafana == nil {
+			grafanaExist = false
+		} else {
+			grafanaExist = true
+		}
+	} else {
+		prometheusExist = false
+		grafanaExist = false
+	}
+
+	// sync prometheus key
+	if prometheusExist {
+		v, err := buildPrometheusEtcdValue(tm)
+		if err != nil {
+			klog.Error(err.Error())
+			return err
+		}
+		err = putPrometheusKey(pdEtcdClient, v)
 		if err != nil {
-			if errors.IsNotFound(err) {
-				tc.Status.Monitor = nil
-			} else {
-				return err
-			}
+			klog.Error(err.Error())
+			return err
+		}
+	} else {
+		err = cleanPrometheusKey(pdEtcdClient)
+		if err != nil {
+			klog.Error(err.Error())
+			return err
+		}
+	}
+
+	// sync grafana key
+	if grafanaExist {
+		v, err := buildGrafanaEtcdValue(tm)
+		if err != nil {
+			klog.Error(err.Error())
+			return err
+		}
+		err = putGrafanaKey(pdEtcdClient, v)
+		if err != nil {
+			klog.Error(err.Error())
+			return err
+		}
+	} else {
+		err = cleanGrafanaKey(pdEtcdClient)
+		if err != nil {
+			klog.Error(err.Error())
+			return err
 		}
 	}
 	return nil
 }
 
+func putGrafanaKey(etcdClient pdapi.PDEtcdClient, value string) error {
```

Comment:
What's the benefit of wrapping up the etcdClient functions here?
The functions are just with different names only.

Reply:
Just want to make things clear and isolated.

Reply:
updated.